File: 2022/day04/main.py
import re

ans, ans2 = 0, 0
r = re.compile(r"^(\d+)-(\d+),(\d+)-(\d+)$")


def is_contains(a, b, c, d):
    if a >= c and b <= d or c >= a and d <= b:
        return True
    else:
        return False


def is_contains2(a, b, c, d):
    if c >= a and c <= b:
        return True
    elif a >= c and a <= d:
        return True
    elif b == c or b == d:
        return True
    else:
        return False


with open("input.txt") as f:
    for line in f:
        line = line.strip()
        m = r.search(line)
        a, b, c, d = [int(m) for m in m.groups()]
        ans += is_contains(a, b, c, d)
        ans2 += is_contains2(a, b, c, d)

print(f"Part 1: {ans}")
print(f"Part 2: {ans2}")

File: 2022/day04/main.go
package main

import (
	"bufio"
	"fmt"
	"os"
)

func IsContains(a, b, c, d uint16) bool {
	if a >= c && b <= d || c >= a && d <= b {
		return true
	} else {
		return false
	}
}

func IsContains2(a, b, c, d uint16) bool {
	if c >= a && c <= b {
		return true
	} else if a >= c && a <= d {
		return true
	} else if b == c || b == d {
		return true
	} else {
		return false
	}
}

func main() {
	var ans, ans2 uint16 = 0, 0
	f, _ := os.Open("input.txt")
	defer f.Close()
	scanner := bufio.NewScanner(f)
	for scanner.Scan() {
		var a, b, c, d uint16
		fmt.Sscanf(scanner.Text(), "%d-%d,%d-%d", &a, &b, &c, &d)
		if IsContains(a, b, c, d) {
			ans++
		}
		if IsContains2(a, b, c, d) {
			ans2++
		}
	}
	fmt.Println("Part 1:", ans)
	fmt.Println("Part 2:", ans2)
}
